pycbc/inference/models/hierarchical.py

```diff
@@ -599,3 +605,227 @@ def _loglikelihood(self):
+            i = numpy.where(parameter_set==parameters)[0][0]
+            self.other_models.update(**parameters)
+            sh_space[i], hh_space[i] = self.other_models.loglr(
+                                            just_sh_hh=True)
+
+        sh_total = sh_ground + sh_space
+        hh_total = hh_ground + hh_space
+        loglr = self.primary_model.marginalize_loglr(sh_total, hh_total)
+        return loglr
+
+    def _loglikelihood(self):
+        other_models_lognl = 0
+        for lbl, model in self.submodels.items():
+            # Update the parameters of each
+            model.update(**{p.subname: self.current_params[p.fullname]
+                            for p in self.param_map[lbl]})
+            other_models_lognl += model.lognl
+
+        # calculate the combined loglikelihood
+        logl = self._loglr() + self.primary_model.lognl + other_models_lognl
+
+        # store any extra stats from the submodels
+        for lbl, model in self.submodels.items():
+            mstats = model.current_stats
+            for stat in self.extra_stats_map[lbl]:
+                setattr(self._current_stats, stat, mstats[stat.subname])
+        return logl
+
+    @classmethod
+    def from_config(cls, cp, **kwargs):
+        r"""Initializes an instance of this class from the given config file.
+        For more details, see `from_config` in `HierarchicalModel`.
+
+        Parameters
+        ----------
+        cp : WorkflowConfigParser
+            Config file parser to read.
+        \**kwargs :
+            All additional keyword arguments are passed to the class. Any
+            provided keyword will override what is in the config file.
+        """
+        # we need the read from config function from the init; to prevent
+        # circular imports, we import it here
+        from pycbc.inference.models import read_from_config
+        # get the submodels
+        submodel_lbls = shlex.split(cp.get('model', 'submodels'))
+        # sort parameters by model
+        vparam_map = map_params(hpiter(cp.options('variable_params'),
+                                       submodel_lbls))
+        sparam_map = map_params(hpiter(cp.options('static_params'),
+                                       submodel_lbls))
+
+        # we'll need any waveform transforms for the initializing sub-models,
+        # as the underlying models will receive the output of those transforms
+        if any(cp.get_subsections('waveform_transforms')):
+            waveform_transforms = transforms.read_transforms_from_config(
+                cp, 'waveform_transforms')
+            wfoutputs = set.union(*[t.outputs
+                                    for t in waveform_transforms])
+            wfparam_map = map_params(hpiter(wfoutputs, submodel_lbls))
+        else:
+            wfparam_map = {lbl: [] for lbl in submodel_lbls}
+        # initialize the models
+        submodels = {}
+        logging.info("Loading submodels")
+        for lbl in submodel_lbls:
```

Comment:
Set primarly model explicitly in the config file. 

[mdoel]
....
primary_model = ce 
other_models = model2 model3